pkg/counter/cache.go

```diff
@@ -0,0 +1,166 @@
+package counter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var cachedCounts = cache{
+	Counts: Counts{
+		UserAccountsPerClusterCounts: map[string]int{},
+	},
+}
+
+// Counts is type that contains number of MURs and number of UserAccounts per member cluster
+type Counts struct {
+	MasterUserRecordCount        int
+	UserAccountsPerClusterCounts map[string]int
+}
+
+type cache struct {
+	Counts
+	sync.RWMutex
+	initialized bool
+}
+
+func write(operation func()) {
+	cachedCounts.Lock()
+	defer cachedCounts.Unlock()
+	operation()
+}
+
+// Reset resets the cached counter - is supposed to be used only in tests
+func Reset() {
+	write(func() {
+		reset()
+	})
+}
+
+func reset() {
+	cachedCounts.Counts = Counts{
+		UserAccountsPerClusterCounts: map[string]int{},
+	}
+	cachedCounts.initialized = false
+}
+
+// IncrementMasterUserRecordCount increments the number of MasterUserRecord in the cached counter
+func IncrementMasterUserRecordCount() {
+	write(func() {
+		cachedCounts.MasterUserRecordCount++
+	})
+}
+
+// DecrementMasterUserRecordCount decreases the number of MasterUserRecord in the cached counter
+func DecrementMasterUserRecordCount(log logr.Logger) {
+	write(func() {
+		if cachedCounts.MasterUserRecordCount != 0 || !cachedCounts.initialized {
```

Comment:
shouldn't it be 
```suggestion
		if cachedCounts.MasterUserRecordCount != 0 && cachedCounts.initialized {
```
instead?

Reply:
nope, my version is correct. 
I don't want to let the counter decrement below zero if the cache is initialized, but I can let go into negative numbers if the counter is not initialized. This cover a small tiny use-case when MUR would be deleted before it is initialized by the content of the ToolchainStatus

Reply:
I'm still not sure why we have to cover such a tiny gap during the startup. I mean, when the counters are initialized, shouldn't we have the exact/absolute values?

Reply:
Probably because I'm too paranoid :man_shrugging: 
I was thinking about this use case:
1. there is some bug why users are not properly removed, or the host-operator pod is dead and meantime some users should be deleted
2. we deploy a new version of host operator
3. at the very beginning, the MUR controller goes through the MUR to be deleted and decrements the counter
4. the ToolchainStatus controller synchronizes with the cache

Reply:
what if we synchronize the cached counters with the database on a regular basis (ie, using some requeuing or some go rountine)? In that case, even if we miss a creation or deletion, the counter would only remain "off by 1" for a certain time, not until the next deployment.

Reply:
> what if we synchronize the cached counters with the database on a regular basis (ie, using some requeuing or some go rountine)?

all the possible approaches were mentioned and discussed in the related doc - this was one of the options
https://docs.google.com/document/d/1W--tiikHRFiiQ3LHiWfY9OXytP8klmT3Xk2JvFSA26k/edit#heading=h.dm34odbbr6ln
but I guess that listing all 3k or 10k MURs every 5 seconds is not very optimal. To avoid such listing I came with the cached counter. 

> In that case, even if we miss a creation or deletion, the counter would only remain "off by 1" for a certain time, not until the next deployment.

just to be clear, the current implementation doesn't synchronize with etcd after each deployment - it lists all existing MURs only once when this feature is deployed to the running cluster. The reason is that we need to count the already provisioned users.
One of the options I mentioned in the doc was that we would synchronize the counter with etcd every 24 hours or something like that. We (me and @alexeykazakov) decided not to do that right now and introduce the synchronization when we figure out that it's needed.